transports/quic/src/transport.rs

```diff
@@ -494,10 +562,18 @@ impl<P: Provider> Stream for Listener<P> {
             match self.new_connections_rx.poll_next_unpin(cx) {
                 Poll::Ready(Some(connection)) => {
                     let local_addr = socketaddr_to_multiaddr(connection.local_addr(), self.version);
-                    let send_back_addr =
-                        socketaddr_to_multiaddr(&connection.remote_addr(), self.version);
+                    let remote_addr = connection.remote_addr();
```

Comment:
I discussed this with @mxinden in today's [open maintainer call](https://github.com/libp2p/rust-libp2p/discussions/4021).

Entries in our hole-punch map are very short-lived. Thus, I'd guess that  an incoming connection from `remote_addr` will very likely resolve to the peer that we are expecting a hole-punch from.

Under this assumption, we could just leave this code here as is and instead look at every event produced by all listeners in `Transport`. For each `TransportEvent::Incoming`, we check if the incoming connection corresponds to a hole-punch and if yes, we take the `Connecting` from the event and send it via the one-shot to the `dial_as_listener` future instead.

This would allow us to remove the "fake" error reporting and I think we can also remove the synchronization around the `HashMap` then? Let me know what you think.

Reply:
If I understand correctly, we have 3 options here:
1. "Fake" error solution at the cost of the library user having to special case their logging of listener errors.
2. Assume the probability of mismatch between an inbound connection and an outgoing hole punch to be 0 (like the [initial version of this PR](https://github.com/libp2p/rust-libp2p/pull/3964/commits/7a714bf57fc5c07267a3e5b07d8ceff7c579a153)).
3. Let `Listener` spawn a separate task for incoming connections at the cost of increased complexity and possible duplication of code with the swarm `Pool`.

I prefer 1 > 2 > 3 for the sake of correctness and lower complexity.

Any thoughts @kpp?

Reply:
> we are behind NAT and thus cannot receive inbound connections at all

> Entries in our hole-punch map are very short-lived.

These are assumptions I cannot process at the moment. Maybe I should join your next open maintainer call to discuss them?